File: go/toolerror/toolerror.go
// Package toolerror defines the shared structured error contract for Demarch MCP servers.
//
// All Demarch MCP tool handlers should return ToolError instead of flat error strings.
// This enables agents to distinguish transient from permanent failures and make
// informed retry decisions.
//
// Usage:
//
//	return toolerror.New(toolerror.ErrNotFound, "agent %q not registered", agentName)
//	return toolerror.New(toolerror.ErrTransient, "database busy").WithRecoverable(true)
package toolerror

import (
	"encoding/json"
	"errors"
	"fmt"
)

// Error type constants. Use these as the Type field in ToolError.
const (
	ErrNotFound   = "NOT_FOUND"   // Resource doesn't exist
	ErrConflict   = "CONFLICT"    // Concurrent modification conflict
	ErrValidation = "VALIDATION"  // Invalid input or arguments
	ErrPermission = "PERMISSION"  // Access denied
	ErrTransient  = "TRANSIENT"   // Temporary failure, safe to retry
	ErrInternal   = "INTERNAL"    // Unexpected server error
)

// ToolError is a structured error for MCP tool handlers.
// It carries enough context for agents to make retry and fallback decisions.
type ToolError struct {
	Type        string         `json:"type"`
	Message     string         `json:"message"`
	Recoverable bool           `json:"recoverable"`
	Data        map[string]any `json:"data,omitempty"`
}

// New creates a ToolError with the given type and formatted message.
// Recoverable defaults to true for ErrTransient, false for all others.
func New(errType string, format string, args ...any) *ToolError {
	return &ToolError{
		Type:        errType,
		Message:     fmt.Sprintf(format, args...),
		Recoverable: errType == ErrTransient,
	}
}

// Error implements the error interface.
func (e *ToolError) Error() string {
	return fmt.Sprintf("[%s] %s", e.Type, e.Message)
}

// WithRecoverable sets the Recoverable flag and returns the error for chaining.
func (e *ToolError) WithRecoverable(r bool) *ToolError {
	e.Recoverable = r
	return e
}

// WithData sets the Data field and returns the error for chaining.
func (e *ToolError) WithData(data map[string]any) *ToolError {
	e.Data = data
	return e
}

// JSON returns the error as a JSON string for embedding in MCP tool results.
func (e *ToolError) JSON() string {
	b, err := json.Marshal(e)
	if err != nil {
		return fmt.Sprintf(`{"type":"%s","message":"%s","recoverable":%t}`, e.Type, e.Message, e.Recoverable)
	}
	return string(b)
}

// FromError extracts a *ToolError from an error chain.
// Returns nil if the error is not a ToolError.
func FromError(err error) *ToolError {
	var te *ToolError
	if errors.As(err, &te) {
		return te
	}
	return nil
}

// Wrap converts any error into a ToolError. If the error is already a ToolError,
// it is returned as-is. Otherwise, it's wrapped as ErrInternal.
func Wrap(err error) *ToolError {
	if te := FromError(err); te != nil {
		return te
	}
	return New(ErrInternal, "%s", err.Error())
}

File: python/interbase/toolerror.py
"""Structured MCP error contract — wire-format compatible with Go toolerror."""

from __future__ import annotations

import json
from typing import Any

# Error type constants — must match Go wire values exactly.
ERR_NOT_FOUND = "NOT_FOUND"
ERR_CONFLICT = "CONFLICT"
ERR_VALIDATION = "VALIDATION"
ERR_PERMISSION = "PERMISSION"
ERR_TRANSIENT = "TRANSIENT"
ERR_INTERNAL = "INTERNAL"

_DEFAULT_RECOVERABLE = {
    ERR_NOT_FOUND: False,
    ERR_CONFLICT: False,
    ERR_VALIDATION: False,
    ERR_PERMISSION: False,
    ERR_TRANSIENT: True,
    ERR_INTERNAL: False,
}


class ToolError(Exception):
    """Structured error for MCP tool handlers.

    Carries enough context for agents to make retry and fallback decisions.
    Wire format matches Go's toolerror.ToolError exactly.
    """

    def __init__(self, err_type: str, message: str, **data: Any) -> None:
        super().__init__(message)
        self.type: str = err_type
        self.message: str = message
        self.recoverable: bool = _DEFAULT_RECOVERABLE.get(err_type, False)
        self.data: dict[str, Any] = dict(data) if data else {}

    def with_recoverable(self, recoverable: bool) -> ToolError:
        """Override the recoverable flag. Returns self for chaining."""
        self.recoverable = recoverable
        return self

    def with_data(self, **kwargs: Any) -> ToolError:
        """Set data fields. Returns self for chaining."""
        self.data.update(kwargs)
        return self

    def json(self) -> str:
        """Serialize to JSON wire format matching Go's encoding/json output."""
        obj: dict[str, Any] = {
            "type": self.type,
            "message": self.message,
            "recoverable": self.recoverable,
        }
        if self.data:
            obj["data"] = self.data
        return json.dumps(obj, separators=(",", ":"))

    def __str__(self) -> str:
        return f"[{self.type}] {self.message}"

    @classmethod
    def from_error(cls, exc: BaseException) -> ToolError | None:
        """Extract a ToolError from an exception. Returns None if not one."""
        if isinstance(exc, ToolError):
            return exc
        return None

    @classmethod
    def wrap(cls, exc: BaseException) -> ToolError:
        """Convert any exception to ToolError. Passthrough if already one."""
        if isinstance(exc, ToolError):
            return exc
        return cls(ERR_INTERNAL, str(exc))
